### DISCUSSION TITLE
Request: option for NVDA to not speak in response when Global Media keys are pressed

### DISCUSSION THREAD

**Discussion by TechHorseG** [nvaccess/nvda#13445] [Ideas]
Hello. When I press the Global Media keys on my keyboard on my Win10 PC a pop-up overlay will appear on screen and NVDA can read out its contents. An example would be where I press the Global Volume Up button and NVDA announces, for instance, "Volume level 70".

My request is simply for an option to disable these announcements. I would like to be able to press any of the Global Media Keys (volume up / down / unmute, Play / Pause etc) without NVDA speaking in response.

NVDA already offers options to disable the automatic reading out of other kinds of pop-up information such as tooltips and tray notifications. I would like to see a similar option for the Global Media keys to be ignored.

NVDA already offers options to disable the automatic reading out of other kinds of keys such as the alphanumeric keys and control keys. I would like to see a similar option for the Global Media keys to be ignored.

That's it, hope this suggestion is OK. The rest of this post explains why I would greatly appreciate this option and can be ignored if you like.

Basically, if NVDA speaks when I press Global Media keys it means I cannot use them on the fly while playing media, as the media will be talked over.

Adjusting the volume, for example, becomes an operation where first I have to pause the content, change the volume guessing at where I want it, resume the media, and then repeat the process if I think more adjustment is needed.

Either that, or I would have to leave the audiobook or whatever playing, adjust the volume to where I want it, then rewind the media a few seconds (not always easy to do) so I can now listen to the last part properly without the "volume level 56", "volume level 58" announcements etc.

Similarly, resuming paused media also becomes an operation where if I press the Global Play / Pause key to resume, then I'd have to try and hit ctrl as quick as I can to stop NVDA talking over the media, and / or rewind a few  secs to listen properly to the audio that was initially missed.

I want to avoid these situations by stopping NVDA from saying anything when I use the Global Media keys, so I can use them on the fly. Thanks for reading.

**Comment by josephsl** [comment]
Hi, hmmm, this is interesting. Tell us more please (such as the app you are using, what NVDA says versus what you are expecting, etc.) so people can get a better idea as to how they can help. Thanks.

**Comment by TechHorseG** [comment]
Hi, an example would be where I press the Global Volume Up button and NVDA announces, for instance, "Volume level 70". I would like to be able to press any of the Global Media Keys (volume up / down / unmute, Play / Pause etc) without NVDA speaking. Thanks for feedback will see if I can edit OP to add this example.

**Comment by josephsl** [comment]
Hi, okay, now we can see things a bit more clearly. Can you tell us more, including which app this is occurring, Windows version you have, and NVDA version you are using? That way we can get to the bottom of this a bit more effectively. Thanks.

**Comment by josephsl** [comment]
Hi,

Don't worry about the Windows version - I see that you have Windows 10. The remaining puzzle pieces are NVDA version, exact steps to reproduce (I have an idea as to how, but please tell us more), app version, and NVDA version. If you are describing what I think is happening, then this will require an add-on to resolve this, but until we have more info, we won't be able to proceed.

Thanks.

**Comment by TechHorseG** [comment]
OK let's see I am using NVDA Version: 2020.4 portable. If for example I press win+M to focus the desktop, then press the global media volume up / down keys, then NVDA announces the current volume level for each adjustment. This means that any media I have playing will be talked over. Netflix, a Youtube video, music playing in any app, anything. The volume announcements will talk over it meaning that I can't just raise or lower the system volume on the fly while listening to or watching  some playing media. Hope that is enough info? Thanks for responses.

**Comment by josephsl** [comment]
Hi,

Ah, we now have the complete picture (to tell you the truth, I had an idea as to what you were talking about, but I asked you to provide more info not only to confirm what I thought, but also to suggest to fill out the issue template completely).

In simple terms, you hear "volume level something" when you press volume keys while all apps are minimized because these messages come from the app that is responsible for displaying the desktop interface. You can resolve this by pressing the volume keys while focused on the app that is playing something - that is, do not minimize all apps before pressing volume keys.

Technical: see #12043 as to why (I consider this a subtle duplicate of the linked issue as volume announcements come from UIA notification event fired by File Explorer). CC @leonardder

Also, while we are at it, NVDA 2020.4 is an old version - I advise upgrading to 2021.3.3 for security unless you are using add-ons that are not updated to support 2021.x features.

Thanks.

**Comment by TechHorseG** [comment]
josephsl, but sometimes it is useful to be able to have media apps playing in the background while not in focus.

IF NVDA is never meant to speak in response to these keys but some bug does, in some circumstances, cause it to, then fair enough, and I hope this bug gets fixed.

Or IF NVDA is actually meant to respond to the pressing of these keys, but some bug is causing it not to in certain circumstances, then I would still request a user option to disable this response at NVDA's end.

Either way, NVDA NEVER speaking in response to global media key presses under any circumstances, is my hoped for option!

Thanks for your time and information, and I will check out that link.

**Comment by josephsl** [comment]
Hi,

I understand your request. For now you can resolve this by:

1. Upgrading to NVDA 2021.3.3. This is a must because of the second step.
2. Download and install Windows App Essentials add-on.
3. After installing Windows App Essentials add-on, restart NVDA, and turn off "report notifications" setting from NVDA menu/Preferences/Settings/Object presentation. This will result in NVDA not announcing ANY notification, including volume keys and toast notifications. Alternatively, you can create an app-specific configuration profile for explorer (File Explorer), then turn off "report notifications" while explorer profile is active.

Note that Narrator will exhibit the same behavior you have described - when you press volume keys, Narrator, too will announce the volume level. This is actually the intended design and NVDA also follows this principle except it will limit announcements to notifications coming from the app you are using.

A while back I wrote a pull request that will actually accomplish what you are asking: create a setting that will let you turn off volume announcements and other announcements (these are called UIA notifications). It was (and still is) stalled due to the settings interface and experience, coupled with life priorities. It might be possible that I might return to it (or someone might pick it up), but in the meantime the solution I noted above will work at the cost of not hearing toast notifications.

@LeonardDer, I think it would be time to rethink about UIA notification event handler strategy. Given recent developments with UIA notification event handling from Windows Terminal, I think it would make sense to tie notification announcement to content changes setting at some point in the future, or at least create the accessible notification event setting discussed a while back.

Thanks.

**Comment by cary-rowen** [comment]
hello

Can we have more configurable options for volume notifications?
For example, there can be play a short media sound, nvda voice synthesizer prompts, and complete silence.
Or depending on the context, if nvda is saying all, then no prompt, or just play a short media sound.

thanks

**Comment by josephsl** [comment]
Hi, playing a sound to replace a UIA notification announcement is not that easy. Not only we need to look at which notifications to replace, but remember that Windows 11 is constantly changing with Insider Preview builds. Although technically and theoretically possible, I don’t think it is truly possible at this time. Thanks.

**Comment by TechHorseG** [comment]
josephsl, again thanks for suggestions, and I understand about the time and some of the technical issues.

Just to say, by all means offer beeps as an alternative to speech, but I don't think I could use them personally. My keyboard doesn't have volume buttons it has a wheel, and I can spin it quite fast when wanting to quickly increase volume. And I wouldn't want to hear six beeps overlapping eachother. Even if I slowed down, not sure I would want to hear beep beep beep, if just wanting to make a quick on-the-fly volume adjustment while watching a video.

All I can ask is that no changes are made which would mean that some audio feedback must now be heard everywhere. The best case scenario is a silence everywhere option if possible.

But if that is not possible and it comes down to a choice between the status quo of speech in Explorer but silence everywhere else, and a new change of having to have at least a beep feedback everywhere, then my vote is definately for the status quo!

**Comment by josephsl** [comment]
Hi, I see. Note that there are times when UIA notification events provide crucial information such as page load status in Microsoft Edge, Calculator results, and soon, terminal output in Windows Terminal (coming soon at least on Windows 11). Therefore, I think the best compromise among the options presented is announce all notifications regardless of where you are or no notifications with an option in between to let NVDA announce notifications (such as volume changes) from the focused app, which is the strategy I outlined in an earlier pull request. Thanks.

**Comment by TechHorseG** [comment]
@josephsl
>"Note that there are times when UIA notification events provide crucial information such as page load status in Microsoft Edge, Calculator results, and soon, terminal output in Windows Terminal (coming soon at least on Windows 11).">

Oh  dear that last part sounds scary as I use Batch Files to run custom timers and background monitors. As in, they will timeout for a custom amount of time then play a timer sound or perform some other task. Last thing I would want is to hear a constant "waiting for X seconds" alert coming from a minimised console whatever I am doing as that would make them unusable. Really hope this isn't on the cards!

>"Therefore, I think the best compromise among the options presented is announce all notifications regardless of where you are or no notifications with an option in between to let NVDA announce notifications (such as volume changes) from the focused app, which is the strategy I outlined in an earlier pull request.">

Perhaps I am misunderstanding but does that last option mean that a focused media app would see NVDA talk if the volume is changed?

One thought. A lot of the beef on this and other threads is with the volume  announcements specifically, rather than all alerts in general. Couldn't NVDA be taught to recognise and ignore these  volume alerts in particular, irrespective of how it handles alerts generally? As I say, just a thought.

**Comment by josephsl** [comment]
Hi,

The option I mentioned (the in-between) is the status quo at the moment precisely in order to silence volume keys from everywhere except in the very place where it comes from: File Explorer. Since File Explorer is responsible for showing the desktop, taskbar and other elements on screen, it is always active unless it crashes. Part of UIA notification event handler inside NVDA checks whether notifications such as volume changes and page loads come from the app you are using at the moment, and if not, notifications will be silenced. This is why I advised not minimizing apps when pressing media keys precisely because when you minimize all apps, system focus moves to desktop, which is ultimately powered by File Explorer.

As for media aps detecting NVDA, no they will not unless the application actively looks for screen reader active flag which NVDA sets when it starts. Since volume notifications come from not from the media app but from File Explorer, as long as you are using the media app or any other app that is not File Explorer, NVDA will not announce volume level changes at all (see above as to why).

As for silencing media keys notification altogether, it is becoming likely that this is not practical:

1. Notification event component changes: one of the components used to handle UIA notifications such as volume level changes is activity Id, a string that denotes what the notification is about. There are two activity Id's I know of for volume announcement: one used in Windows 10, and a different Id used in Windows 11 (preview if I recall correctly). Because of the nature of Windows as a Service (WaaS), or for that matter, given that dev channel builds will serve as a space for early testing, it is likely that Microsoft will change activity Id's from time to time. This rules out one solution: silencing media keys by checking for a specific set of activity Id's. There is another component that's responsible for notifications known as "display string", and silencing NVDA based on display string isn't even on the table (Windows localization is the reason).
2. Multimodal confirmation: in recent Windows 11 Insider Preview builds, when you change volume level, a visual confirmation is also shown. This means what UIA notification event reports (such as NVDA announcing volume level changes) reflects what sighted users will actually see (the new approach is a small tooltip-like popup). Therefore, from both practical and philosophical point of view, it actually will not make sense to silence volume keys when focused on File Explorer (I add philosophy here because of the goal and mission of screen reading, which is to report relevant screen information to users).

Given what I wrote, I think it makes sense to think about a notification event setting in NVDA or keep the status quo. As for the overall UIA notification event handling approach, I think NVDA uses a better strategy (only announced notifications coming from the focused app), although I do understand the need to announce all notifications regardless of where you are. I also realize that what I wrote might be a bit hard to digest - feel free to read it many times and think about it; again I'm open to possibilities.

Thanks.

**Comment by TechHorseG** [comment]
josephsl thanks that is all very clear. Now that I know that the volume key announcements are all contained to only occur in focussed explorer-based applications and that this is done deliberately, I feel a bit more relaxed as I know how to guarantee avoiding them interupting my listening to media. Whatever NVDA decides to try in the future, then you certainly can't go wrong with allowing people the option to keep the current behaviour, as a safety net. I have no further questions or queries, thanks again.

**Comment by XLTechie** [comment]
@TechHorseG You can silence these announcements with a dictionary entry, if you really want them gone that badly.

* Enter your default dictionary: NVDA+n, p, d, d
* Add a new entry: alt+a
* In the pattern field write: `\bvolume level \d+\b`
* Put nothing in the replacement field.
* In the comment field write something like: Silence volume announcements
* Leave case sensitive unchecked.
* For the type, choose "regular expression": press the up arrow once to reach it.
* Press OK.
* Press OK again.

**Comment by seanbudd** [comment]
Please move this discussion to a [GitHub discussion](https://github.com/nvaccess/nvda/discussions/new).
Once concrete feature requests are ready, please create new issues filling out the template correctly.

**Comment by feerrenrut** [comment]
Converted issue into a discussion.
